File: data-structures/linked-lists/linked-lists.ts
class LinkedListNode<T> {
  constructor(public value: T, public next?: LinkedListNode<T>) {}

  toString = () => {
    let curr: LinkedListNode<T> | undefined = this
    let output = ''
    do {
      if (curr?.value) output += String(curr?.value)
      curr = curr?.next
    } while (curr)

    return output
  }
}

const arrayToLinkedList = <T>(arr: T[]) => {
  let head: LinkedListNode<T> | undefined
  let curr: LinkedListNode<T>

  for (const n of arr) {
    const newNode = new LinkedListNode(n)
    if (!head) {
      head = newNode
    } else {
      curr!.next = newNode
    }
    curr = newNode
  }

  return head
}

const ll = new LinkedListNode(1, new LinkedListNode(2, new LinkedListNode(3)))
console.log(ll.toString())
console.log(String(arrayToLinkedList([1, 2, 3, 4, 5])).toString())

File: data-structures/linked-lists/linked-lists.py
class Node(object):
  def __init__(self, value, next=None):
    self.value = value
    self.next = next

  def __str__(self):
    curr = self
    output = ''
    while curr:
      output += str(curr.value)
      curr = curr.next
    return output
  
def listToLinkedList(nums):
  head = None
  curr = None
  for n in nums:
    new_node = Node(n)
    if not head:
      head = new_node
    else:
      curr.next = new_node
    curr = new_node
  return head
  

ll = Node(1, Node(2, Node(3)))
print(ll)
print(listToLinkedList([1, 2, 3, 4, 5]))
